File: examples/crud-item-list/src/main/java/com/gurunars/crud_item_list/example/ActivityMain.java
package com.gurunars.crud_item_list.example;

import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import com.gurunars.crud_item_list.CrudItemList;
import com.gurunars.crud_item_list.ItemEditListener;
import com.gurunars.crud_item_list.ListChangeListener;

import java.util.List;

import butterknife.ButterKnife;


public class ActivityMain extends AppCompatActivity {

    private CrudItemList<AnimalItem> crudItemList;
    private Model model;
    private View creationMenu;

    private void initData(boolean force) {
        if (model.wasInited() && !force) {
            crudItemList.setItems(model.getItems());
            return;
        }
        model.clear();
        for (int i=0; i < 1; i++) {
            model.createItem(new AnimalItem(AnimalItem.Type.LION));
            model.createItem(new AnimalItem(AnimalItem.Type.TIGER));
            model.createItem(new AnimalItem(AnimalItem.Type.MONKEY));
            model.createItem(new AnimalItem(AnimalItem.Type.WOLF));
        }

        crudItemList.setItems(model.getItems());
    }

    private void confItemType(@IdRes int id, final AnimalItem.Type type) {
        creationMenu.findViewById(id).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                model.createItem(new AnimalItem(type));
                crudItemList.setItems(model.getItems());
            }
        });
        crudItemList.registerItemType(type, new AnimalRowBinder());
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        model = new Model(this);
        setContentView(R.layout.crud_list);

        creationMenu = View.inflate(this, R.layout.create_layout, null);

        crudItemList = ButterKnife.findById(this, R.id.customView);

        crudItemList.setListChangeListener(new ListChangeListener<AnimalItem>() {
            @Override
            public void onChange(List<AnimalItem> items) {
                for (AnimalItem item: items) {
                    model.updateItem(item);
                }
            }

        });
        crudItemList.setEmptyViewBinder(new EmptyBinder());
        crudItemList.setCreationMenu(creationMenu);
        crudItemList.setItemEditListener(new ItemEditListener<AnimalItem>() {
            @Override
            public void onEdit(AnimalItem editableItem) {
                editableItem.update();
                model.updateItem(editableItem);
                crudItemList.setItems(model.getItems());
            }

        });

        confItemType(R.id.lion, AnimalItem.Type.LION);
        confItemType(R.id.tiger, AnimalItem.Type.TIGER);
        confItemType(R.id.monkey, AnimalItem.Type.MONKEY);
        confItemType(R.id.wolf, AnimalItem.Type.WOLF);

        initData(false);

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.activity_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int i = item.getItemId();
        switch (i) {
            case R.id.leftHanded:
                crudItemList.setLeftHanded(true);
                break;
            case R.id.rightHanded:
                crudItemList.setLeftHanded(false);
                break;
            case R.id.reset:
                initData(true);
                break;
            case R.id.lock:
                setTitle(R.string.unsortable);
                crudItemList.setSortable(false);
                break;
            case R.id.unlock:
                setTitle(R.string.sortable);
                crudItemList.setSortable(true);
                break;
            case R.id.addMany:
                addMany();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void addMany() {
        model.clear();
        for (int i=0; i < 20; i++) {
            model.createItem(new AnimalItem(AnimalItem.Type.LION));
            model.createItem(new AnimalItem(AnimalItem.Type.TIGER));
            model.createItem(new AnimalItem(AnimalItem.Type.MONKEY));
            model.createItem(new AnimalItem(AnimalItem.Type.WOLF));
        }
        crudItemList.setItems(model.getItems());
    }

}

File: examples/crud-item-list/src/main/kotlin/com/gurunars/crud_item_list/example/ActivityMain.kt
package com.gurunars.crud_item_list.example

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import com.gurunars.animal_item.AnimalItem
import com.gurunars.animal_item.Service
import com.gurunars.animal_item.Service.Companion.getRealService
import com.gurunars.box.box
import com.gurunars.box.patch
import com.gurunars.box.ui.layoutAsOne
import com.gurunars.box.ui.decorators.statefulLayer
import com.gurunars.crud_item_list.IconColorBundle
import com.gurunars.crud_item_list.ItemTypeDescriptor
import com.gurunars.crud_item_list.crudItemListView

class ActivityMain : Activity() {

    private lateinit var srv: Service

    private val isSortable = true.box

    private fun getType(i: Int): AnimalItem.Type {
        return if (isSortable.get()) {
            when (i % 4) {
                0 -> AnimalItem.Type.LION
                1 -> AnimalItem.Type.TIGER
                2 -> AnimalItem.Type.MONKEY
                else -> AnimalItem.Type.WOLF
            }
        } else {
            AnimalItem.Type.MONKEY
        }
    }

    private fun addItems(
        count: Int
    ) {
        srv.items.patch {
            val id = (this.map { it.id }.lastOrNull() ?: 0) + 1
            this + (0 until count).map {
                AnimalItem(
                    id = id + it,
                    type = getType(it),
                    version = 0
                )
            }
        }
    }

    private fun reset() {
        srv.clear().subscribe {
            addItems(4)
        }
        val mngr = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        mngr.primaryClip = ClipData.newPlainText("", "")
    }

    private fun initView(sortable: Boolean) {
        setTitle(if (sortable) R.string.sortable else R.string.unsortable)

        val descriptors: List<ItemTypeDescriptor<AnimalItem>>

        if (sortable) {
            descriptors = listOf(
                R.drawable.ic_menu_monkey to AnimalItem.Type.MONKEY,
                R.drawable.ic_menu_lion to AnimalItem.Type.LION,
                R.drawable.ic_menu_tiger to AnimalItem.Type.TIGER,
                R.drawable.ic_menu_wolf to AnimalItem.Type.WOLF
            ).map {
                Descriptor(
                    srv.items,
                    this,
                    it.first,
                    it.second
                )
            }
        } else {
            descriptors = listOf(Descriptor(
                srv.items,
                this,
                R.drawable.ic_menu_monkey,
                AnimalItem.Type.MONKEY)
            )
        }

        statefulLayer(R.id.main) {
            setBackgroundColor(Color.WHITE)
            retain(isSortable)
            crudItemListView(
                confirmationActionColors = IconColorBundle(
                    bgColor = Color.BLACK,
                    fgColor = Color.GREEN
                ),
                cancelActionColors = IconColorBundle(
                    bgColor = Color.RED,
                    fgColor = Color.WHITE
                ),
                openIconColors = IconColorBundle(
                    bgColor = Color.GREEN,
                    fgColor = Color.YELLOW
                ),
                items = srv.items,
                itemTypeDescriptors = descriptors,
                sortable = sortable,
                addToTail = sortable,
                actionIconColors = IconColorBundle(
                    fgColor = Color.YELLOW,
                    bgColor = Color.BLUE
                ),
                clipboardSerializer = Serializer()
            ).layoutAsOne(this)
        }.layoutAsOne(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        srv = getRealService()
        isSortable.onChange(listener = this::initView)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.activity_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val i = item.itemId
        when (i) {
            R.id.reset -> reset()
            R.id.lock -> {
                isSortable.set(false)
                isSortable.broadcast()
                reset()
            }
            R.id.unlock -> {
                isSortable.set(true)
                isSortable.broadcast()
                reset()
            }
            R.id.addMany -> addItems(4 * 20)
        }
        return super.onOptionsItemSelected(item)
    }
}
